lib/binary/binary.nit

```diff
@@ -0,0 +1,303 @@
+		return [for b in 8.times do int.bin_and(2**b) > 0]
+	end
+
+	# Read a floating point on 32 bits and return it as a `Float`
+	#
+	# Using this format may result in a loss of precision as it uses less bits
+	# than Nit `Float`.
+	fun read_float: Float
+	do
+		if last_error != null then return 0.0
+
+		var b0 = read_byte
+		var b1 = read_byte
+		var b2 = read_byte
+		var b3 = read_byte
+
+		# Check for error, `last_error` is set by `read_byte`
+		if b0 == null or b1 == null or b2 == null or b3 == null then return 0.0
+
+		return native_read_float(b0, b1, b2, b3)
+	end
+
+	# Utility for `read_float`
+	private fun native_read_float(b0, b1, b2, b3: Int): Float `{
+		union {
+			unsigned char b[4];
+			float f;
+		} u;
+
+		u.b[0] = b0;
+		u.b[1] = b1;
+		u.b[2] = b2;
+		u.b[3] = b3;
+
+		return u.f;
+	`}
+
+	# Read a floating point on 64 bits and return it as a `Float`
+	fun read_double: Float
+	do
+		if last_error != null then return 0.0
+
+		var b0 = read_byte
+		var b1 = read_byte
+		var b2 = read_byte
+		var b3 = read_byte
+		var b4 = read_byte
+		var b5 = read_byte
+		var b6 = read_byte
+		var b7 = read_byte
+
+		# Check for error, `last_error` is set by `read_byte`
+		if b0 == null or b1 == null or b2 == null or b3 == null or
+		   b4 == null or b5 == null or b6 == null or b7 == null then return 0.0
+
+		return native_read_double(b0, b1, b2, b3, b4, b5, b6, b7)
+	end
+
+	# Utility for `read_double`
+	private fun native_read_double(b0, b1, b2, b3, b4, b5, b6, b7: Int): Float `{
+		union {
+			unsigned char b[8];
+			double f;
+		} u;
+
+		u.b[0] = b0;
+		u.b[1] = b1;
+		u.b[2] = b2;
+		u.b[3] = b3;
+		u.b[4] = b4;
+		u.b[5] = b5;
+		u.b[6] = b6;
+		u.b[7] = b7;
+
+		return u.f;
+	`}
+
+	# Read a signed integer on 64 bits and return is an `Int`
+	#
+	# Using this format may result in a loss of precision as the length of a
+	# Nit `Int` may be less than 64 bits on some platforms.
+	fun read_int64: Int
+	do
+		if last_error != null then return 0
+
+		var b0 = read_byte
+		var b1 = read_byte
+		var b2 = read_byte
+		var b3 = read_byte
+		var b4 = read_byte
+		var b5 = read_byte
+		var b6 = read_byte
+		var b7 = read_byte
+
+		# Check for error, `last_error` is set by `read_byte`
+		if b0 == null or b1 == null or b2 == null or b3 == null or
+		   b4 == null or b5 == null or b6 == null or b7 == null then return 0
+
+		return native_read_int64(b0, b1, b2, b3, b4, b5, b6, b7, big_endian)
+	end
+
+	# Utility for `read_int64`
+	private fun native_read_int64(b0, b1, b2, b3, b4, b5, b6, b7: Int, big_endian: Bool): Int `{
+		union {
+			unsigned char b[8];
+			int64_t val;
+			uint64_t conv;
+		} u;
+
+		u.b[0] = b0;
+		u.b[1] = b1;
+		u.b[2] = b2;
+		u.b[3] = b3;
+		u.b[4] = b4;
+		u.b[5] = b5;
+		u.b[6] = b6;
+		u.b[7] = b7;
+
+		if (big_endian)
+			u.conv = be64toh(u.conv);
+		else u.conv = le64toh(u.conv);
+
+		return u.val;
+	`}
+end
+
+redef class Int
+	# Utility for `BinaryWriter`
+	private fun int64_byte_at(index: Int, big_endian: Bool): Int `{
```

Comment:
We could use a `int32_byte_at` too I guess, since the Nit Int has no length warranties, it could be a 32 bit value instead of a 64.

That, or the `int64_byte_at` could handle the case in C via a `sizeof(long)` or something like that, then we could pad with 0 for the 32 most significant bits.